File: Stack/C++/739_daily_temperature.cpp
/*
* RUNTIME: 37 ms - Beats 12.28% of C++ submissions
* MEMORY - 112.00 MB - Beats 05.03% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
#include <set>
#include <stack>
#include <unordered_map>
#include <functional>
using namespace std;


// My thought process:
//
// 1. Initialize a result vector with the same length as `temperatures` and set all elements to 0.
// 2. Initialize a stack that stores pairs of (temperature, index).
// 3. Loop through the `temperatures` array using an index variable to track both day and temperature.
// 4. While the stack is not empty and the current temperature is higher than the top element’s temperature:
//      - Pop the last element from the stack.
// 5. Calculate the number of days it took to reach a higher temperature:
//      - result[previous_index] = current_day - previous_index;
// 6. If the current temperature is not higher, push (temperature, index) onto the stack and continue.
// 7. After iterating through all temperatures, return the result vector.


class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> res(n, 0);
        stack<pair<int, int>> st;

        for (int curr_day = 0; curr_day < n; ++curr_day)
        {
            int temp = temperatures[curr_day];
            while (!st.empty() && temp > st.top().first)
            {
                int prev_temp = st.top().first;
                int prev_day = st.top().second;
                st.pop();
                res[prev_day] = curr_day - prev_day;
            }
            st.push({temp, curr_day});
        }

        return res;
    }
};

File: Stack/Python/739_daily_temperature.py
# RUNTIME: 121 ms, faster than 40.42% of python3 Submissions. 
# MEMORY USAGE: 33.80 MB, less than 16.81% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Initiate a restultstack with len(temperature) elements and set all to "0"
# 2. Initaite a stack that contains a touple containing the temperature and the index
# 3. loop through as enumerate to get the index value aswell
# 4. while stack is available and new temperature is higher pop last stack element
# 5. input the days it took to get the higher value into resultstack
#   5.1 to get the amount of days take current day  - stack index (inital day)
# 6. if temperature is lower append to st and iterate through next for loop cycle
# 7. if all elements of "temperatures" are cycled return the result list


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        st = []

        for curr_day, temp in enumerate(temperatures):
            while st and temp > st[-1][0]:
                _, st_ind = st.pop()
                res[st_ind] = curr_day - st_ind

            st.append((temp, curr_day))

        return res
    

ls  = [73,74,75,71,69,72,76,73] 
sol = Solution()
print (sol.dailyTemperatures(ls))